EIPS/eip-3337.md

```diff
@@ -0,0 +1,102 @@
+
+| Opcode    | Gas used |
+|-----------|----------|
+| `PUSHn x` | 3        |
+| `MSTOREFP`| 3        |
+
+Consuming only 6 gas and at least 3 bytes of bytecode, while once again only requiring EVM implementations to do one extra addition operation. The alternative of storing values on the stack requires 6 gas and 2 bytes of bytecode for the sequence `SWAPn POP`, making it no more efficient than memory storage.
+
+## Specification
+### Parameters
+
+| Constant | Value |
+| - | - |
+| `FORK_BLOCK` | TBD |
+
+For blocks where `block.number >= FORK_BLOCK`, the following changes apply.
+
+### Frame pointer
+A new EVM internal state variable called the "frame pointer" is introduced. This is a signed integer that starts at 0.
+
+### `SETFP` opcode
+A new opcode, `SETFP` is introduced with value `0x5c`. This opcode costs `G_low` (3 gas) and takes one argument from the stack. The argument is stored as the new value of the frame pointer.
+
+### `GETFP` opcode
+A new opcode, `GETFP` is introduced with value `0x5d`. This opcode costs `G_low` (3 gas) and takes no arguments. It takes the current value of the frame pointer and pushes it to the stack.
+
+### `MLOADFP` opcode
+A new opcode `MLOADFP` is introduced with value `0x5e`. This opcode acts in all ways identical to `MLOAD`, except that the value of the frame pointer is added to the address before loading data from memory. An attempt to load data from a negative address should be treated identically to an invalid opcode, consuming all gas and reverting the current execution context.
+
+### `MSTOREFP` opcode
+A new opcode `MSTOREFP` is introduced with value `0x5f`. This opcode acts in all ways identical to `MSTORE`, except that the value of the frame pointer is added to the address before storing data to memory. An attempt to store data to a negative address should be treated identically to an invalid opcode, consuming all gas and reverting the current execution context.
+
+## Rationale
+### Cost of new opcodes
+The cost of the new opcodes `MLOADFP` and `MSTOREFP` reflects the cost of `MLOAD` and `MSTORE`. They are generally equivalent in cost with the exception of an extra addition operation, which imposes negligible cost.
+
+The cost of the new opcodes `SETFP` and `GETFP` is based on other common low-cost opcodes such as `PUSH` and `POP`.
+
+### Absence of `MSTORE8FP`
+Whether or not to include an `MSTORE8FP` opcode depends on the frequency with which this will be used, and the desire to minimise the size of the instruction set and to conserve opcodes for future use. If the community believes this will be a substantial asset, the EIP can be modified to include it.
+
+## Backwards Compatibility
+This EIP exclusively introduces new opcodes, and as a result should not impact any existing programs unless they operate under the assumption that these opcodes are undefined, which we believe will not be the case.
+
+## Test Cases
+TBD
+
```

Comment:
```suggestion
```
Not a required section, so I recommend just leaving it out until such time in the future as you fill it in.  If you want to leave this in as a reminder to yourself, that is also acceptable.